### PR TITLE
Use ZMQ sockets to communicate results of pytest run

### LINKED ISSUE #112: Plugin confused by tests writing to sys.__stdout__
For instance, a test file containing

```python
import sys
def test_foo():
    print('Barf 3', file=sys.__stdout__)
```

leads to the traceback included below. I guess the solution is for the communication between the pytestworker and the plugin to go over a socket, as Spyder does with introspection, instead over stdout.

### Traceback

```
  File "/home/jitse/misc/anaconda/lib/python3.6/site-packages/spyder_unittest/backend/jsonstream.py", line 93, in consume
    len_encoding = int(txt[index:end_of_line1])
ValueError: invalid literal for int() with base 10: 'Barf 3'

Traceback (most recent call last):
  File "/home/jitse/misc/anaconda/lib/python3.6/site-packages/spyder_unittest/backend/pytestrunner.py", line 44, in read_output
    result = self.reader.consume(output)
  File "/home/jitse/misc/anaconda/lib/python3.6/site-packages/spyder_unittest/backend/jsonstream.py", line 96, in consume
    % (repr(txt), index, end_of_line1))
ValueError: txt = 'Barf 3\n188\n{"event": "logreport", "when": "call", "outcome": "passed", "nodeid": "test_fast.py::test_foo", "sections": [], "duration": 0.0005002021789550781, "filename": "test_fast.py", "lineno": 41}\n632\n{"event": "finished", "stdout": "============================= test session starts ==============================\\nplatform linux -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0\\nPyQt5 5.6 -- Qt runtime 5.6.2 -- Qt compiled 5.6.2\\nrootdir: /home/jitse/misc/testunittest, inifile:\\nplugins: xdist-1.15.0, qt-2.2.1, mock-1.6.3, json-0.4.0, cov-2.4.0, hypothesis-3.38.5, flaky-3.4.0\\ncollected 1 item\\n\\ntest_fast.py .                                                           [100%]\\n===Flaky Test Report===\\n\\n\\n===End Flaky Test Report===\\n\\n=========================== 1 passed in 0.01 seconds ===========================\\n"}\n'  index = 0  end_of_line1 = 6
```

### PR DESCRIPTION
**Pull request by jitseniesen**
Communicating the results of the pytest run via stdout, as is done now, is fragile. The tests itself may also write to stdout and stderr and it was not clear to me how to block this off completely in complicated cases such as the Spyder tests, where the Qt C library writes to stdout. Communicating over a socket, as implemented here, seems to be more robust. It's also (arguably) cleaner.

Fixes #112.

### CODE REVIEW

**Comment by ccordoba12** [comment]
You should add pyzmq to `setup.py`, just in case Spyder stops using it someday.

File: setup.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def get_package_data(name, extlist):
 
 
 # Requirements
-REQUIREMENTS = ['lxml', 'spyder>=3']
+REQUIREMENTS = ['lxml', 'spyder>=3', 'zmq']
```
Comment (ccordoba12) [inline setup.py]: The Python package is called `pyzmq`, not zmq

File: conda.recipe/meta.yaml [hunk 2]
```diff
@@ -25,6 +25,7 @@ requirements:
     - lxml
     - python
     - spyder >=3
+    - zmq
```
Comment (ccordoba12) [inline conda.recipe/meta.yaml]: pyzmq

File: circle.yml [hunk 3]
```diff
@@ -13,14 +13,14 @@ machine:
     COVERALLS_REPO_TOKEN: c5Qt1n27dLFCAIESYVkuCmVpUU8doney1
     # Environment variables used by astropy helpers
     TRAVIS_OS_NAME: "linux"
-    CONDA_DEPENDENCIES: "lxml mock nose pytest"
+    CONDA_DEPENDENCIES: "lxml mock nose pytest zmq"
```
Comment (ccordoba12) [inline circle.yml]: pyzmq

File: appveyor.yml [hunk 4]
```diff
@@ -18,7 +18,7 @@ environment:
     PYTHON_ARCH: "64" # needs to be set for CMD_IN_ENV to succeed. If a mix
                       # of 32 bit and 64 bit builds are needed, move this
                       # to the matrix section.
-    CONDA_DEPENDENCIES: "lxml mock nose pytest spyder"
+    CONDA_DEPENDENCIES: "lxml mock nose pytest spyder zmq"
```
Comment (ccordoba12) [inline appveyor.yml]: pyzmq

File: .travis.yml [hunk 5]
```diff
@@ -17,7 +17,7 @@ env:
     - TEST_CI="True"
     - TEST_CI_APP="True"
     # Environment variables used by astropy helpers
-    - CONDA_DEPENDENCIES="lxml mock nose pytest spyder"
+    - CONDA_DEPENDENCIES="lxml mock nose pytest spyder zmq"
```
Comment (ccordoba12) [inline .travis.yml]: pyzmq

**Comment by jitseniesen** [comment]
@ccordoba12 Thanks! Note to self: Always test or suffer embarrassment.

**Comment by ccordoba12** [comment]
> Always test or suffer embarrassment.

Jeje, no worries. I'd be drown in brown paper bags for every time I've embarrassed myself in Spyder :-)

**Pull request merged**